### PR TITLE
gh-93554: Conditional jumps only jump forward

### PR DESCRIPTION
**Pull request by iritkatriel**
* Issue: gh-93554

### CODE REVIEW

File: Doc/library/dis.rst [hunk 1]
```diff
@@ -999,60 +999,44 @@ iterations of the loop.
    .. versionadded:: 3.11
 
 
-.. opcode:: POP_JUMP_FORWARD_IF_TRUE (delta)
+.. opcode:: POP_JUMP_IF_TRUE (delta)
 
    If TOS is true, increments the bytecode counter by *delta*.  TOS is popped.
 
    .. versionadded:: 3.11
```
Comment (sweeneyde) [inline Doc/library/dis.rst]: Should `versionadded` be removed? I think py3.10 and before had `POP_JUMP_IF_TRUE`, but it always did `CHECK_EVAL_BREAKER()` in case it was backward.

Comment (iritkatriel) [reply Doc/library/dis.rst]: It was an absolute jump in 3.10. I guess this could be versionchanged, to say it's relative as of 3.11. But this PR is not going to be back ported, so it would need to be a different PR that can be back ported to 3.11.

Comment (markshannon) [inline Doc/library/dis.rst]: `POP_JUMP_IF_TRUE` has been around forever, so this should probably be removed. Maybe add a note that it was temporarily removed in 3.11.

File: Doc/library/dis.rst [hunk 2]
```diff
@@ -999,60 +999,44 @@ iterations of the loop.
...
+   .. versionchanged:: 3.12
+      This is no longer a pseudo-instruction.
 
 
-.. opcode:: POP_JUMP_FORWARD_IF_FALSE (delta)
+.. opcode:: POP_JUMP_IF_FALSE (delta)
 
    If TOS is false, increments the bytecode counter by *delta*.  TOS is popped.
 
    .. versionadded:: 3.11
```
Comment (markshannon) [inline Doc/library/dis.rst]: Same as above.

File: Python/compile.c [hunk 3]
```diff
@@ -7738,63 +7718,91 @@ assemble_emit(struct assembler *a, struct instr *i)
...
+            break;
+        case POP_JUMP_IF_FALSE:
+            reversed_opcode = POP_JUMP_IF_TRUE;
+            break;
+        case POP_JUMP_IF_TRUE:
+            reversed_opcode = POP_JUMP_IF_FALSE;
+            break;
+        case JUMP_IF_TRUE_OR_POP:
+        case JUMP_IF_FALSE_OR_POP:
+            if (!is_forward) {
```
Comment (markshannon) [inline Python/compile.c:7763]: We should handle this case, rather than hoping it won't happen.

We can test for it now, thanks to your new test framework. Can't we?

Or do you plan to do that in another PR?

Comment (iritkatriel) [reply Python/compile.c:7763]: I can do it in this PR, but let's first finish the test framework, i.e. merge https://github.com/python/cpython/pull/96274 .

Comment (markshannon) [reply Python/compile.c:7763]: :+1:

**Review by brandtbucher — APPROVED** [review]
Thanks for doing this!

Just curious, have you benchmarked this change?

**Comment by brandtbucher** [comment]
We might also consider adding a test that loops ending in branches can still quicken code.

For example, something like:

```py
def f(x: bool) -> None:
    for i in range(1_000_000):
        if x:
            pass
```

Today `f(True)` quickens the code, but `f(False)` doesn't. After this change, both should.

**Comment by iritkatriel** [comment]
> Just curious, have you benchmarked this change?

I haven't yet.

File: Python/compile.c (hunk 3 above)
Comment (iritkatriel) [reply Python/compile.c:7763]: Actually let's do that separately. This PR is large enough and that is a different story.

**Comment by markshannon** [comment]
[Small speedup[(https://gist.github.com/markshannon/3250fe8d47d7f3532633d39024d0cb43)

**Comment by iritkatriel** [comment]
> We might also consider adding a test that loops ending in branches can still quicken code.
> 
> For example, something like:
> 
> ```python
> def f(x: bool) -> None:
>     for i in range(1_000_000):
>         if x:
>             pass
> ```
> 
> Today `f(True)` quickens the code, but `f(False)` doesn't. After this change, both should.
> 
> Ditto for `while` loops. This code should be quickened after only one call:
> 
> ```python
> def w() -> None:
>     i = 0
>     while i < 1_000_000:
>         i += 1
> ```

Where does such a test belong?

**Comment by iritkatriel** [comment]
I'll merge this while the buildbots are happy with it, we can add more tests (for specialisation etc) in a followup PR.
